Make app movable to SD card

Atrate opened the pull request:
## Rationale
Over 25% of the world's Android devices still run Android 6.0 or older [[1]](https://www.androidauthority.com/android-version-distribution-748439/). This version of Android still has native support for moving applications to SD cards — this is crucial when dealing with low-storage devices. Many more recent OEM OSes also support moving applications to SD cards. This means making applications movable to the SD card is IMHO a rational choice.

## Solution
Add `android:installLocation="auto"` flag to AndroidMainfest.xml. The default for Android is, curiously, not `"auto"` but `"internalOnly"`

https://developer.android.com/guide/topics/data/install-location

Signed-off-by: Atrate <Atrate@protonmail.com>

Atrate committed:
Make app movable to SD card

Signed-off-by: Atrate <Atrate@protonmail.com>

Commit:
Nullable Address.domainId

domainId for Address.kt and Address.java must be nullable, differently from what explained in the doc
Affected: AddressesBridgeMapper.kt

Commit:
Changed build version

#comment Just increased the build version code and name for a new beta
Affected: nothing

Commit:
Changed build version

Affected: nothing

Commit:
Merge branch 'develop' into prerelease

# Conflicts:
#	buildSrc/src/main/kotlin/ProtonMail.kt

Commit:
Send button is disabled in draft message

MAILAND-1025

Commit:
Merge branch 'fix/1025_send-button-disabled-in-draft-message' into 'prerelease'

Send button disabled in draft message

See merge request android/mail/proton-mail-android!203

Commit:
Changed build version

#comment Just increased the build version code for a new beta
Affected: nothing

Commit:
Make the detekt analysis job not mandatory for release branches

If detekt fails we will still push to github as we can build.

Commit:
Merge branch 'feat/make-detekt-analysis-not-mandatory-for-release' into 'release'

Make the detekt analysis job not mandatory for release branches

See merge request android/mail/proton-mail-android!205

Commit:
Make the detekt analysis job not mandatory for tags

Commit:
Merge branch 'feat/make-detekt-analysis-not-mandatory-for-tags' into 'release'

Make the detekt analysis job not mandatory for tags

See merge request android/mail/proton-mail-android!207

Commit:
Make manual release the first stage

Commit:
Merge branch 'feat/make-manual-release-the-first-stage' into 'release'

Make manual release the first stage

See merge request android/mail/proton-mail-android!208

Commit:
Merge branch 'release' into prerelease

# Conflicts:
#	buildSrc/src/main/kotlin/ProtonMail.kt

Commit:
Merge branch 'develop' into prerelease

# Conflicts:
#	buildSrc/src/main/kotlin/ProtonMail.kt

Commit:
App crashes on Logout with Android 11

MAILAND-1045

#comment Affected: Logging out of the app no matter the Android version..

Commit:
Merge branch 'fix/1045_app-crashes-on-logout-with-Android11' into 'prerelease'

App crashes on Logout with Android 11

See merge request android/mail/proton-mail-android!223

Commit:
Added queries to aid package visibility handling

MAILAND-1047,
MAILAND-1050

#comment With API level 30 there are changes with package visibility with required adding queries in Manifest in order the app to handle them correctly. Affected: Opening external apps from inside the app, like ex. uploading image, clicking on telephone number to open dialer, opening links etc..

Commit:
Merge branch 'fix/1047_1050_api-level-30-package-visibility-handled' into 'prerelease'

Added queries to aid package visibility handling

See merge request android/mail/proton-mail-android!224

gtomek committed:
Made database instances created by Dagger/Hilt in DatabaseModule depend on username in order to support multi-user feature correctly.

MAILAND-1052

gtomek committed:
Put retry to FetchContactsDataWorker as it used to be in FetchContactsJob and put default empty string to @CurrentUsername in ApplicationModule to fix multilpe accounts not appearing scenario.

MAILAND-1052

gtomek committed:
Removed retry from LogoutWorker as it was causing delayed logout issue in multi user configuration due to delayed logout worker retry with invalid auth token.

MAILAND-1052

Commit:
Merge branch 'fix/1052_mixed-contact-multi-user' into 'prerelease'

Made database instances created by Dagger/Hilt in DatabaseModule depend on...

See merge request android/mail/proton-mail-android!225

Commit:
Changed build version

#comment Just increased the build version code for a new beta
Affected: nothing

Commit:
Merge branch 'develop' into prerelease

# Conflicts:
#	buildSrc/src/main/kotlin/ProtonMail.kt

Commit:
Merge branch 'develop' into prerelease

gtomek committed:
Make repost bug section extend BaseActivity instead of BaseConnectivityActivity.

MAILAND-1099

Commit:
Changed build version

#comment Just increased the build version code for a new beta build
Affected: nothing

gtomek committed:
Crash in NetworkConnectivityManager on Marshmallow and below fixed.

#comment The used method ConnectivityManager#registerDefaultNetworkCallback() has been added in API24, therefore
any Android version before API24 (Android Marshmallow) didn't have access to this method and the app was crashing.
It has been replaced with ConnectivityManager#registerNetworkCallback() method which is available from API21 (the min
version supported by the application).

MAILAND-1106

Commit:
Merge branch 'fix/1106_connectivity-manager-crash-on-marshmallow-and-below-prerelease' into 'prerelease'

Crash in NetworkConnectivityManager on Marshmallow and below fixed.

See merge request android/mail/proton-mail-android!240

Commit:
Changed build version

#comment Just increased the build version code and name for a new beta
Affected: nothing

gtomek committed:
Fix crash in CreateAccountFeedbackActivity, removed BaseConnectivityActivity.

MAILAND-1113

Commit:
Changed build version

#comment Just increased the build version code and name for a new beta
Affected: nothing

Commit:
Merge branch 'develop' into prerelease

# Conflicts:
#	buildSrc/src/main/kotlin/ProtonMail.kt

marinomeneghel committed:
Update displayed contact details when contact is updated

For the reasons explained in 8b0564c68 the ContactsEvents in
ContactDetailsActivity is not being received anymore. To ensure we
display the updated contact details we listen for activity result

MAILAND-1155

Commit:
Merge branch 'fix/1155-update-displayed-contact-details-after-editing' into 'prerelease'

Update displayed contact details after editing

See merge request android/mail/proton-mail-android!254

Commit:
Changed build version

#comment Just increased the build version code for a new beta
Affected: nothing

Commit:
Fixes an dependence issue with testing stage inside ci file

#comment Affected: Running of CI jobs..

StefanijaBoshkovska committed:
Provide deserialize method in ProcessPushNotificationDataWorker with PushNotification serializer

MAILAND-1259

StefanijaBoshkovska committed:
Define priority for new email notifications for devices with API level 25 or below

Commit:
Reformat FetchMessageCountsJob.kt

Fixed code style and simplified logic

MAILAND-1254

StefanijaBoshkovska committed:
Merge branch 'fix/improve-showing-notifications' into 'prerelease'

Improve showing notifications

See merge request android/mail/proton-mail-android!287

Commit:
Add missing `onSwitchedAccounts` from NavigationActivity.kt

MAILAND-1254

StefanijaBoshkovska committed:
Merge branch 'fix/1254_user-switch-logout' into 'prerelease'

[1254] Fix user switch logout from drawer

See merge request android/mail/proton-mail-android!291

Commit:
Changed build version

#comment Just increased the build version code for a new beta
Affected: nothing

Commit:
Merge branch 'develop' into prerelease

# Conflicts:
#	.gitlab-ci.yml
#	buildSrc/src/main/kotlin/ProtonMail.kt

Commit:
Merge branch 'prerelease' into release

# Conflicts:
#	buildSrc/src/main/kotlin/ProtonMail.kt

Commit:
Merge branch 'develop' into prerelease

dimitar10 committed:
Use a blocking api call for refreshing tokens.

Affected: Token refresh logic.

MAILAND-1364

StefanijaBoshkovska committed:
Change build version code

#comment Increased the build version code for a new beta
Affected: nothing

Commit:
Merge branch 'fix/1364_user-not-logged-out-revoke-session-2' into 'prerelease'

Use a blocking api call for refreshing tokens

See merge request android/mail/proton-mail-android!327

StefanijaBoshkovska committed:
Revert "Replace FcmIntentService with ProcessPushNotificationDataWorker"

This reverts commit ccd05619

StefanijaBoshkovska committed:
Merge branch 'fix/revert-replacing-fcm-intent-service-with-worker' into 'prerelease'

Revert "Replace FcmIntentService with ProcessPushNotificationDataWorker"

See merge request android/mail/proton-mail-android!330

marinomeneghel committed:
Revert "Convert message recipient class to kotlin"

This reverts commit ae103b1b16f1e65abdd4bad84e7bcbc0d8383c71.

marinomeneghel committed:
Revert addition of equals to MessageRecipient

This change needs to be reverted in order to fix a deserialization issue
caused by the `serialVersionUID` used for serializing this class in
older versions changes becuase of the addition of equals.

Before doing changes to this class, we need to declare a custom
`serialVersionUID` for it (possibily the one the existing class has) or
plan for a lightweight migration

Also delete MessageRecipientMatcher as unused

MAILAND-1373

marinomeneghel committed:
Manually check equality of MessagePayload objects in ServerMessageTest

This is needed because `assertEquals` will fail since MessageRecipient
doesn't implement equals. Equals was just removed from MessageRecipient
in the previous commit as it caused issues with serialization, reason
for which we're not adding it back right now. MAILAND-1386 is tracking
an improvement on that side.

MAILAND-1373

Commit:
Merge branch 'fix/1373-resolve-undisclosed-receipient-regression-1' into 'prerelease'

Resolve "undisclosed recipient" regression

See merge request android/mail/proton-mail-android!334

Commit:
Changed build version

#comment Just increased the build version code and name for a new beta
Affected: nothing

Commit:
Merge branch 'develop' into prerelease

# Conflicts:
#	app/src/main/java/ch/protonmail/android/api/interceptors/ProtonMailAuthenticator.kt
#	app/src/main/java/ch/protonmail/android/api/models/MessageRecipient.java
#	app/src/main/java/ch/protonmail/android/api/models/room/messages/Message.kt
#	app/src/main/java/ch/protonmail/android/api/utils/Fields.kt
#	app/src/main/java/ch/protonmail/android/contacts/groups/list/ContactGroupsFragment.kt
#	app/src/main/java/ch/protonmail/android/fcm/FcmIntentService.java
#	app/src/main/java/ch/protonmail/android/jobs/UpdateAndPostDraftJob.java
#	app/src/main/java/ch/protonmail/android/servers/notification/NotificationServer.kt
#	app/src/test/java/ch/protonmail/android/api/interceptors/ProtonMailAuthenticatorTest.kt
#	buildSrc/src/main/kotlin/ProtonMail.kt

Commit:
Resolves a conflict. Added back injection for NotificationManager.

Commit:
Merge branch 'fix/1419_app-crash-when-you-open-composer' into 'prerelease'

fix/1373 'Undisclosed Recipients' AND fix/1419 App crash when you open composer

See merge request android/mail/proton-mail-android!348

Commit:
Fixes issue with kotlin serialization by introducing the serialVersionUID in the affected class `MessageRecipient`. And a similar issue with field `Constants.MessageActionType.NONE` by add a rule to proguard.

MAILAND-1373
MAILAND-1419

#comment Affected: Opening of the composer and message details sender field, also contacts..

Commit:
Changed build version

#comment Just increased the build version code for a new beta
Affected: nothing

Commit:
When fetching contacts we were making too many calls to `/contacts/emails` due to wrong paging calculations.

Affected: fetching and refreshing contacts

Commit:
Merge branch 'fix/regression-fetching-contacts-wrong-paging' into 'prerelease'

Fix/fetching contacts wrong paging

See merge request android/mail/proton-mail-android!353

Commit:
Changed build version

#comment Just increased the build version code for a new beta
Affected: nothing

marinomeneghel committed:
Do not update draftId with value returned by AddAttachmentsActivity

Changes:
- Do not update viewmodel's draftId with the one returned by AddAttachmentsActivity
- Observe outcome of draft update (not only creation) from VM (as with
  auto save it's now possible this flow is executed while the composer
  is still open)

This solves a regression found during testing of build 756, for which
following the steps outlined in MAILAND-1300 an invalid, duplicated copy
of the draft would be created locally. The attachments added to the
draft would get lost.

The cause was this bit of logic in ComposeMessageActivity that "updated"
the VM's `draftId` with the value that was initially passed to
AddAttachmentsActivity. Since while the user is on `Attachments` screen
the update draft flow gets triggered by the auto save functionality, if
the user statys in the attachments screen long enough for the update to
complete the ID that we have on VM will be the updated one, which
differs from the local one that Attachments screen will return.

This wasn't happening before the refactor introduced by MAILAND-1281 as
the AddAttachmentsActivity would receive an `Event` through the bus with
the updated ID of the draft (which was then returned)

MAILAND-1300

Commit:
Merge branch 'fix/1300-regression-creating-invalid-local-draft-when-adding-attachments-before-draft-creation-completed' into 'prerelease'

Fix 1300 regression creating invalid local draft when adding attachments before draft creation completed

See merge request android/mail/proton-mail-android!355

Commit:
Changed build version

#comment Just increased the build version code for a new beta
Affected: nothing

marinomeneghel committed:
Upload Public key to the message only when sending it

This solves an issue for which the public key was uploaded as part of
saving a draft, each time it was saved.

This issue was introduced with commit 6319af6bef33122b94266d, which began re-using
the UploadAttachments use case logic without considering that while such logic
does indeed upload attachments, it was extracted from the sending process
(MAILAND-1139) and it contains some op (namely, upload of oublic key)
that are expected to be done only while sending and not when saving a draft.

MAILAND-1511

Commit:
Merge branch 'prerelease' into release

marinomeneghel committed:
Interrupt sending of the message if attachments upload is still in progress

This patches an issue for which if the attachments upload is started but
not ended before a message is sent, the attachments not yet uploaded
won't be sent.

MAILAND-1511

marinomeneghel committed:
Merge branch 'fix/1511-avoid-uploading-public-key-multiple-times' into 'prerelease'

MAILAND-1511 Avoid uploading public key multiple times

See merge request android/mail/proton-mail-android!368

Commit:
Changed build version

#comment Just increased the build version code and name for a new beta
Affected: nothing

gtomek committed:
FetchContactsDataWorker improved handling of CancelableException.

MAILAND-1537

gtomek committed:
Display proton contacts before merging them with android contacts in case user refuses access to android contacts preventing any contacts being displayed.

MAILAND-1537

gtomek committed:
Changed contacts list item types resolution logic to prevent one contact from dissapearing.

MAILAND-1537

gtomek committed:
Some reformatting and logging improvement in contacts section.

MAILAND-1537

marinomeneghel committed:
Format MessageFactory following detekt rules

MAILAND-1555

marinomeneghel committed:
Solve issues for which removed parent attachments were sent of Fw

THis issue was happening because removing parent attachments in the app (for a
message beign forwarded) was not being reflected in the backend.

The underlying cause is that performing a `copy` of the draft's creation
response, together with setting local attachments on the created draft,
was causing the loss of the reference to any attachments that were
created through the create draft request (such as parent attachments).

Consequentially, we did not have the correct ID of the attachment when
trying to perform the deletion, causing the response to be handled
silently (as it's the same case of the attachment not existing remotely) and
the attachments to be sent with the message.

MAILAND-1555

marinomeneghel committed:
Attachments can be removed also before draft creation finishes on FW

This solves a regression for which if a user access the
attachments screen and removes any (parent) attachments before the
initial draft creation finished, those attachments would not be removed
and sent with the message instead.

The underlying cause for this regression is that until the draft is
created, attachments do have a "localId" which is not the same that
backend generates at draft's creation time.
For this reason, the removal fails silently as backend doesn't find any
attachments with such ids.

The solution consists in making AddAttachmentsActivity listen for
changes to the Message object in the DB, displaying the attachments only
after the draft creation finished (or immediately, in case the user is
offline)

MAILAND-1555

marinomeneghel committed:
Cancel pending create draft requests before sending

This is needed to replicate the existing logic of not saving a draft
when sending is ongoing, (performed though the check for pendingForSend
in saveDraft use case) also when the message is sent while offline.

This prevents a race condition between saving the draft and sending
which might cause the message to be sent with attachments that were
instead removed from the offline message.

Moving the pendingForSend check to CreateDraftWorker was also evaluated,
but eventually this solution was preferred as that check was removed in
the send refactor branch which is currently under QA (so this solution
will apply to those changes too)

MAILAND-1555

marinomeneghel committed:
AttachmentsViewModel does not listen for draft updated events

As we only want to delay displaying the attachments when draft is still
being created

MAILAND-1555

marinomeneghel committed:
Fixup! Solve compilation error with ComposeMessageViewModelTest

Commit:
updated css file to address zoom out issue in webview

Commit:
Merge branch 'fix/1540_editor-text-zoomed-out-css' into 'prerelease'

updated css file to address zoom out issue in webview

See merge request android/mail/proton-mail-android!387

marinomeneghel committed:
Pass messageId to AttachmentsViewModel through SavedStateHandle

MAIAND-1555

marinomeneghel committed:
Extract AttachmentsViewState to a separated file

MAILAND-1555

marinomeneghel committed:
Post view state of LiveData instead of setting value

This removes the need to switch on Main thread with `withContext` for
simplicy.

An additional explicit check for the CoroutineScope to be active was
added to ensure additional values posted on the flow are not considered
after the scope was cancelled (which happens once the first
UpdateAttachments event was posted)

MAILAND-1555

marinomeneghel committed:
Simplify flow collection

MAILAND-1555

marinomeneghel committed:
AttachmentsViewModel uses NetworkConnectivityManager to check network

As we deem this component to provide more reliable information compared
to QueueNetworkUtil (as the latter takes into considerations more data
than just whether we do have internet connection or not)

MAILAND-1555

marinomeneghel committed:
Rename draftCreationHappened method for clarity

MAILAND-1555

marinomeneghel committed:
Rename findMessageByDbId to respect naming conventions

MAILAND-1555

marinomeneghel committed:
Merge branch 'fix/1555-removed-attachments-are-being-sent' into 'prerelease'

Fix "removed attachments are being sent on forward"

See merge request android/mail/proton-mail-android!386

Commit:
Changed build version

#comment Just increased the build version code and name for a new beta
Affected: nothing

marinomeneghel committed:
Make findMessageByDbId return optional message into the flow

As this message can actually be null, and it will cause an NPE if not
treated as optional

MAILAND-1555

marinomeneghel committed:
Remove distinctUntilChanged as not needed

Commit:
Merge branch 'fix/1555-handle-null-messages-from-DB' into 'prerelease'

Make findMessageByDbId return optional message into the flow

See merge request android/mail/proton-mail-android!393

Commit:
Changed build version

#comment Just increased the build version code for a new beta
Affected: nothing

marinomeneghel committed:
Solve issue where active VPN plan is lost when subscribing a Mail paid plan

This issue was caused by the API call to `createUpdateSubscription` API
being performed with only the `planId` of the Mail plan the user just
chose, omitting any existing plans the user already has as described
 by the `fetchSubscription` API response.

Technically speaking, said existing plans were being omitted because an
exception was thrown when `CreateSubscription` class was trying to add
the planIds returned by the API to the `planIds` list.

The core of the issue lies in the callers of `CreateSubscription`, which
as `planIds` parameter pass lists that are not mutable.
For example, the call from UpsellingActivity is performed passing
`Collections.singletonList(mSelectedPlanId)` which is fine at compile
time but throws exception at runtime as it's not mutable.

The solution was to change planIds to a mutableList directly in
`CreateSubscription` before performing any operations on it.

Other changes:
- Equals was added to CreateSubscriptionBody POJO for unit testing purposes

MAILAND-1640

marinomeneghel committed:
Solve issue where active VPN plan is lost when subscribing a Mail paid plan

This issue was caused by the API call to `createUpdateSubscription` API
being performed with only the `planId` of the Mail plan the user just
chose, omitting any existing plans the user already has as described
 by the `fetchSubscription` API response.

Technically speaking, said existing plans were being omitted because an
exception was thrown when `CreateSubscription` class was trying to add
the planIds returned by the API to the `planIds` list.

The core of the issue lies in the callers of `CreateSubscription`, which
as `planIds` parameter pass lists that are not mutable.
For example, the call from UpsellingActivity is performed passing
`Collections.singletonList(mSelectedPlanId)` which is fine at compile
time but throws exception at runtime as it's not mutable.

The solution was to change planIds to a mutableList directly in
`CreateSubscription` before performing any operations on it.

Other changes:
- Equals was added to CreateSubscriptionBody POJO for unit testing purposes

MAILAND-1640

marinomeneghel committed:
Merge branch 'fix/1640-solve-issue-where-subscribing-new-mail-plan-cancelled-other-existing-plans' into 'prerelease'

Solve issue where subscribing new mail plan cancelled other existing plans

See merge request android/mail/proton-mail-android!400

Commit:
Changed build version

#comment Just increased the build version code for a new beta
Affected: nothing

Commit:
Merge branch 'fix/1640-solve-issue-where-subscribing-new-mail-plan-cancelled-other-existing-plans-release' into 'release'

Hotfix issue where subscribing new mail plan cancelled other existing plans

See merge request android/mail/proton-mail-android!402

Commit:
fixed CI job that had wrong branch name

Commit:
Changed build version

#comment Just increased the build version code for a release
Affected: nothing

gtomek committed:
Merge branch 'fix/1537_contacts-list-empty-for-prerelease' into 'prerelease'

Fix/1537 contacts list empty for prerelease

See merge request android/mail/proton-mail-android!404

Commit:
Changed build version

#comment Just increased the build version code for a new beta
Affected: nothing

Commit:
Merge branch 'prerelease' into release

# Conflicts:
#	buildSrc/src/main/kotlin/ProtonMail.kt

Commit:
Merge branch 'develop' into prerelease

# Conflicts:
#	app/src/main/java/ch/protonmail/android/activities/messageDetails/repository/MessageDetailsRepository.kt
#	app/src/main/java/ch/protonmail/android/api/models/room/messages/MessagesDatabase.kt
#	app/src/main/java/ch/protonmail/android/api/services/PostMessageServiceFactory.kt
#	app/src/main/java/ch/protonmail/android/attachments/UploadAttachments.kt
#	app/src/main/java/ch/protonmail/android/compose/ComposeMessageViewModel.kt
#	app/src/main/java/ch/protonmail/android/jobs/messages/PostMessageJob.java
#	app/src/main/java/ch/protonmail/android/usecase/compose/SaveDraft.kt
#	app/src/main/java/ch/protonmail/android/worker/drafts/CreateDraftWorker.kt
#	app/src/test/java/ch/protonmail/android/attachments/UploadAttachmentsTest.kt
#	app/src/test/java/ch/protonmail/android/compose/ComposeMessageViewModelTest.kt
#	app/src/test/java/ch/protonmail/android/usecase/compose/SaveDraftTest.kt
#	app/src/test/java/ch/protonmail/android/worker/CreateDraftWorkerTest.kt
#	buildSrc/src/main/kotlin/ProtonMail.kt

Commit:
Crash in Drafts with error when retrieving a recipient.. This wasn't actually connected to the recipient, but to the cookie management library which should be fixed in this new version.

MAILAND-1661

Commit:
fix FileNotFoundException when viewing or downloading attachment

Commit:
Merge branch 'fix/1661_crash-from-cookie-menagment-library' into 'prerelease'

Fix/1661 crash from cookie management library

See merge request android/mail/proton-mail-android!410

Commit:
When we introduced global handling of 422 http error code it resulted with toast showing up in places where we have questionable behaviours that don't really cause issues for the end user. And sadly the errors provided by the API weren't of any particular use to the end user. So for now we show only the toast for this error codes.

MAILAND-1663
MAILAND-1664

Commit:
Merge branch 'fix/file-not-found-exception' into 'prerelease'

fix/FileNotFoundException

See merge request android/mail/proton-mail-android!412

Commit:
Add the location id of the message we are trying to delete as currentLabelID to the endpoint to avoid concurrency issues

MAILAND-1556

Affected: Deleting a message from Mailbox, MessageDetails and Composer

Commit:
revert name of build type for release

marinomeneghel committed:
Re-apply "Save parent attachments which were already uploaded when draft is created"

This reverts commit 4a9e63810cbecca72ead972a5c4e502b14cb66da.

This was removed to replicate the fix introduced for MAILAND-1555 in
release branch. Since the logic that handles drafts / attachments /
sending was changed in develop due to refactoring of the send
functionality (MAILAND-914) this piece of logic that preserves local
attachments doesn't cause the issue detailed in MAILAND-1555 anymore,
and it actually prevents other issues with attachments being out of
sync compared to what the user sees locally.

In particular, this handling local attachments prevents an issue for which
attachments would not be sent when adding them to a FW message
that had already attachments (parent attachments)

MAILAND-1665

marinomeneghel committed:
Send Message fails when saving draft failed all the retries

This is to avoid having nested retry cycles caused by the
sendMessageWorker re-triggering saving draft after all the tries failed
already.

MAILAND-1665

marinomeneghel committed:
Do not explicitly cancel ongoing draft creation when sending message

This was introduced by 0254ae69f1 to solve a race condition between the
ongoing save draft and the sending operation, which was causing
attachments being sent to differ from attachments the user was seeing.

This erroneous behaviour was happening as the logic to create a draft in
"SaveDraft" use case was differing from the one in "PostMessageJob".
Since SendMessageWorker was introduced the logic for saving draft is
exactly the same (SaveDraft use case) at any time (be it saving a draft
or sending). The case of sending a message while a draft being saved
will be handled implicitly from the SendMessageWorker, which  by
re-scheduling the SaveDraft use case will cancel the current job.

- Decrease detekt maxIssues

MAILAND-1665

Commit:
Refine test for 422 response, to check whether toast shows up for users or no, depending on the error code

marinomeneghel committed:
Reduce max retries to 3 for SaveDraft, Send and Upload Attachments

As the `runAttemptCount` begins from 0, the code was allowing 4 retries
to be performed for each of the operations mentioned above.

MAILAND-1665

Commit:
Revert "revert name of build type for release"

This reverts commit 4a256271

Commit:
Merge branch 'fix/specific-error-codes-handling' into 'prerelease'

Fix/Specific error codes handling

See merge request android/mail/proton-mail-android!413

marinomeneghel committed:
SaveDraft use case returns Result instead of flow of Result

This makes collection simpler, as we always only care about one value in
this case.
It also avoids some cases where the collection of the "saveDraft" flow
would block execution of a method and not allow code after the call to
save draft to be executed (eg. usage in ComposeMessageVM was behaving
like this)

MAILAND-1685

marinomeneghel committed:
Fix error sending message when sender was changed multiple times

This error was triggered by some of the attachments being changed more
than twice on a message for which some of the attachments were already
uploaded. The underlying cause, was that "oldSenderId" field on
ComposeMessageVM was not updated when being changed the second time,
causing the re-encryption of the attachment to fail (as we couldn't
decrypt it with the key of the wrong 'oldSender') and consequentially
the send operation to fail.

The fix consisted in always updating the old sender field with the
value of the current sender once the draft was updated. To ensure this
happens timely, we also trigger saving a draft when the sender changes.

MAILAND-1685

marinomeneghel committed:
Solve deserialization issue causing a crash when opening some messages

Usage of kotlin serialization in place of "dedicated" string serialization
logic from AttachmentHeaders class was creating a crash when trying to
deserialize data that was serialized from previous versions
(reproduced with 764 to 765)

MAILAND-1689

gtomek committed:
Some reformatting in MailTo and ComposeMessageActivity.

MAILAND-322

gtomek committed:
New MailToParser added.

MAILAND-322

gtomek committed:
Fixed mailto links parsing with the new version of androidx-core (1.5-rc01) which has also been updated.

MAILAND-322

gtomek committed:
Fixed cursos not appearing in seach bar.

MAILAND-1110

gtomek committed:
Replaced getColor by ContextCompat#getColor to fix crash on API21-22 also reformatted NavigationActivity.

MAILAND-1110

marinomeneghel committed:
Increase global network requests timeout to 30 seconds

and ProtonRetrofitExtended instances timeout to one minute (used in
potentially long running ops such as uoloading attachments)

MAILAND-1687

gtomek committed:
Fix for crashing encrypted messages, reverted AttachmentHeaders kt conversion.

MAILAND-1693

gtomek committed:
Merge branch 'fix/1693_specific-encrypted-messages-causing-a-crash' into 'prerelease'

Solve deserialization issue causing a crash when opening some messages

See merge request android/mail/proton-mail-android!427

marinomeneghel committed:
Merge branch 'fix/1665-attachment-not-sent-if-added-while-fw-message-with-attachments' into 'prerelease'

Fix attachment not sent if added while fw message with attachments

See merge request android/mail/proton-mail-android!417

Commit:
Changed build version

#comment Just increased the build version code for a new beta
Affected: nothing

gtomek committed:
Resize contacts avatar bitmap to viewpoint size before saving it to prevent out of memory problems.

MAILAND-1271

gtomek committed:
Updated android worker and material design versions.

MAILAND-1271

gtomek committed:
Some refactor in EditContactDetails.

MAILAND-1271

gtomek committed:
Pass encrypted vcard file path instad of vcard string to avoid worker arguments size limitations.

MAILAND-1271

gtomek committed:
Updated CreateContactTest and CreateContactWorkerTest added some methods to FileHelper.

MAILAND-1271

Commit:
fixed unit test where currentLabelId was missing

gtomek committed:
Save vcard to a file when starting edit contact flow to prevent big images crashing the app.

MAILAND-1271

gtomek committed:
Simplified one method in FileHelper, updated CreateContactTest.

MAILAND-1271

Commit:
Merge branch 'prerelease' into release

marinomeneghel committed:
Make usage of encrypted contacts data available to all users

MAILAND-1666

StefanijaBoshkovska committed:
Remove translatable false flag from add_contact_details string

The string needs to be translated because it is shown on a button, so this false flag needs to be removed.

MAILAND-1697

StefanijaBoshkovska committed:
Replace array tag with string-array in strings.xml

Array is not recognized on Crowdin and it prevents from these string resources being translated.

MAILAND-1697

Commit:
Add .idea/assetWizardSettings.xml to .gitignore

Commit:
Fetch assert4k from Maven Central

gtomek committed:
Fixed saving multiple times different attachments with the same name, causing a counter being added in the wrong place of file name due to mime type resolution.

MAILAND-1855

darigovresearch commented:
Think this issue would be really helpful if merged. Can anyone advise on what is needed to resolve the conflicts?

Commit:
Merge branch 'fix/cherry-pick-saving-multiple-times-same-attachment' into 'prerelease'

Fixed saving multiple times different attachments with the same name, causing...

See merge request android/mail/proton-mail-android!504

Commit:
Remove donate feature

Commit:
Remove all resources connected to donate

Commit:
Merge branch 'fix/remove-donate-feature' into 'prerelease'

Fix/Remove donate feature

See merge request android/mail/proton-mail-android!534

Commit:
Changed build version

Affected: nothing

Commit:
Merge branch 'prerelease' into release

adamproton committed:
Add "Open in Proton Calendar" button

CALAND-899

adamproton committed:
Open Play Store on button click if Proton Calendar is not installed

adamproton committed:
Force opening ICS attachments with Proton Calendar when clicking on CTA

gtomek committed:
Use standard detekt plugin locally instead of cutom proton plugin.

MAILAND-1792

gtomek committed:
Updated detekt config, set counter at current value 6885.

MAILAND-1792

gtomek committed:
Updated gitlab build script to use standard detekt plugin for analysis.

MAILAND-1792

gtomek committed:
Updated MessageRenderer and related tests to match what is aready on develop.

MAILAND-1792

gtomek committed:
Fix for GetSavedCredentialsTest.

MAILAND-1792

gtomek committed:
Merge branch 'chore/1792_detekt-use-standard-local-plugin' into 'prerelease'

Use standard detekt plugin locally instead of cutom proton plugin.

See merge request android/mail/proton-mail-android!540

gtomek committed:
Create new file if it does not exist before saving attachment in downloads before API29

MAILAND-1970

gtomek committed:
Retry downloading attachemtns 3 times.

MAILAND-1970

gtomek committed:
Updated ComposeMessageViewModelTest.

MAILAND-1970

adamproton committed:
Don't show the button if Proton Calendar is installed but can't handle ICS

neilmarietta committed:
Upgraded GOpenPGP to 2.1.9.

gtomek committed:
Merge branch 'fix/1970-improve-attachement-download-prerelease' into 'prerelease'

Attachment downloads changes

See merge request android/mail/proton-mail-android!555

Commit:
Add logging Throwable.message to SentryTree.kt

Commit:
Add usage of obfuscateEmails to SentryTree.kt

Commit:
Add DetailedException.kt for log more details to Sentry

Commit:
Improve logs for Sentry Issue #4K2QW

Commit:
Moved the showing of the snackbar for `server not reachable` only after the ping returns an error

Commit:
Improve DetailedException.kt to automatically include error code and message if available

Commit:
Improve CreateDraftWorker.kt logging

Commit:
Adapt SendMessageWorkerTest.kt to new logging

MAILAND-2001

neilmarietta committed:
Upgraded GOpenPGP to 2.1.10.

Commit:
Improve SendMessageWorker.kt TODO comment with Jira ticket

MAILAND-2001

Commit:
Add test for DetailedException.kt in Timber.DebugTree

MAILAND-2001

adamproton committed:
Open ICS files only with custom Proton Calendar intent action, pass email addresses as extras

Commit:
Add test for DetailedException.kt in SentryTree.kt

MAILAND-2001

Commit:
Add messageId to Failure result for CreateDraftWorker.kt

MAILAND-2001

Commit:
Handle CancellationException in SendMessageWorker.kt

MAILAND-2001

Commit:
Ignore DetailedExceptionTimberDebugTest.kt

Test are running correctly but fails on CI for unknown reasons. We spent enough time trying to understand ( including trying to run them with different Java versions ), without any result

adamproton committed:
Handle Calendar Button with OnAttachmentDownloadCallback, so we request storage permission if it's not granted yet

CALAND-1115

Commit:
The "server not reachable" banner was blinking and showing up multiple times. Because when some specific exceptions were received from the server we were assuming that server is not reachable, right away.

This commit introduces a new connectivity state, that is emitted in this situations, when we a call to the api resulted with an exception. In that case we are not sure if the server is reachable, so the state is called PING_NEEDED and after that we ping the server and only show the banner once the ping returns a result.

MAILAND-1807

Commit:
Rename .java to .kt

Commit:
Merge branch 'chore/2046_gopenpgp-2110' into 'prerelease'

Prerelease GoOpenPGP 2.1.10

See merge request android/mail/proton-mail-android!582

Commit:
Fixes test for class FetchContactsData, they were returning `getSeconds` method not found

Commit:
Merge branch 'fix/cherry-picked-from-develop-multiple-fixes' into 'prerelease'

Cherry picks from develop for multiple fixes

See merge request android/mail/proton-mail-android!569

Commit:
Changed build version

#comment Just increased the build version code for a new beta
Affected: nothing

Commit:
Merge branch 'prerelease' into release

marinomeneghel committed:
Do not treat "Message Already Sent" as an error when sending a message

This is to avoid some cases where errors would be displayed to users
when there was no actual issue.
One such case might happen when killing the app right after sending a
message:
If the "send message" API request was already fired but the worker
didn't complete yet, when the worker will be restarted by the OS it
will enter a failure loop as the save draft use case will fail with
"Message Already sent" error.

At this point, we handle the error by propagating it up to the send
worker, which will ensure the given message is moved out of
drafts to the send folder and then will log this error and finish.

MAILAND-2010

Commit:
Fix draft body erased on app opening after low memory

When the app was close by the system, for example on low memory, after reopen it, it was treated in the same was as when a new message has being composed, thus the message body field was being setup with an empty message + signature.
Now, beside checking if the extras contains the message id as extra, we check if it contains the draft it as state

MAILAND-1989

marinomeneghel committed:
Merge branch 'feat/open-in-calendar' into 'prerelease'

Add "Open in Proton Calendar" button

See merge request android/mail/proton-mail-android!538

marinomeneghel committed:
Solve issue when forwarding an old message from search list

Sending was failing when trying to forward a message from search list.
This was happening when the message was old enought to not be available
in the messagesDatabase anymore, and was caused by the fact that sending
process doesn't take into account the "searchDatabase" when searching
for the parent message (as such DB is scheduled for removal).

The solution is to fetch the parent message from the API when not
available locally.

- Removing the `parentMessage?.attachments(database)` method call as not
  needed, since the same is performed by
  `messageDetailsRepository.findMessageByIdBlocking()`. This was a
  leftover from the old logic which was performing this additional call
  as it was choosing which DB to pass in dynamically (messages / search)

MAILAND-2010

marinomeneghel committed:
Refresh displayed composer attachments when opening AddAttachmentsActivity

after a draft was created.

This is needed to solve the issue detailed in MAILAND-2010/comment-84445:
When forwarding a message from search, if the attachments screen was
opened after the draft creation happened it would contain outdated
attachmentIds, causing the removal of attachments to fail.

This fix ensures that the displayed attachments represent what is
available in the DB in this moment.

MAILAND-2010

Commit:
Merge branch 'fix/1989_prerelease-erase-body-after-app-close' into 'prerelease'

[1989] PRERELEASE - Fix draft body erased on app opening after low memory

See merge request android/mail/proton-mail-android!620

marinomeneghel committed:
Merge branch 'fix/2010_improve-sending-flow-error-handling' into 'prerelease'

Do not treat "Message Already Sent" as an error when sending a message

See merge request android/mail/proton-mail-android!616

Commit:
Changed build version

#comment Just increased the build version code and name for a new beta
Affected: nothing

Commit:
When all alternative proxies would fail the only way to retry was by clicking the `Retry` button on the red `Server not reachable banner`. In rare cases the flow would emit the CANT_REACH_SERVER state and try to show the banner while doh is still trying proxies, but we don't want to show the banner until doh is done, to avoid false positives. And in this case the user didn't have a way of resolving the networking, since the `Retry` button on the banner and the banner is not there..

Now we check connectivity on mailbox refresh and as a result of that retry the proxies if needed. But we also force the banner to show if all proxies have failed.

marinomeneghel committed:
Rename onDohFailed callback and call super from MailboxActivity

MAILAND-2218

Commit:
Merge branch 'fix/doh-retry-improvements-when-all-proxies-fail-v3' into 'prerelease'

[ prerelease ] Improve ways to restore networking when all proxies fail

See merge request android/mail/proton-mail-android!639

Commit:
Changed build version

#comment Just increased the build version code and name for a new beta
Affected: nothing

Commit:
Merge branch 'prerelease' into release

Commit:
Changed build version

#comment Just increased the build version code and name for a new release
Affected: nothing

marinomeneghel committed:
Reduce time between "send message" operation retries

When any step of the "Send message" operation fails, an exponential
backoff policy was applied. In the worst case scenario (each of the
possible steps of the sending process failing each possibile
attempt) this "retry policy" would imply long delays and cause the
send operation to become too long (and the user to lose the context of
the ongoing operation).

This commit reduces the number of retries for the "save draft" and
"upload attachments" operations (both part of the sending process) to 1,
the one for "send message" to 2 and the backoff delay to 10 seconds for
all ops.

MAILAND-2293

marinomeneghel committed:
Avoid cancellation of Send message API request and its handling

This is to prevent cases of the worker being cancelled after the API
request was fired and before the response was handled, which might leave
the user with some inconsistent state (eg. the message might be sent but
not be moved in the sent folder locally)

MAILAND-2293

marinomeneghel committed:
Updated .gitignore to exclude intellij's deploymentTargetDropDown

marinomeneghel committed:
Send Message worker finds message by messageId when dbId fails

This is to add a layer or resieliency to a potential failure case that
would happen if a "replace" operation in Room causes this message's DB
ID to change.

MAILAND-2293

marinomeneghel committed:
Merge branch 'fix/2293_sending-improvements' into 'prerelease'

Improve send message process reliability

See merge request android/mail/proton-mail-android!678

marinomeneghel committed:
Changed build version

ZoricaStojchevska committed:
Merge branch 'prerelease' into release

Atrate committed:
Merge branch 'release' into feature/sd

marinomeneghel commented:
We will be temporarily closing our repository for collaboration while we work on rewriting the Proton Mail Android app. 

During this time, this repository won’t accept any issues or pull requests. We will be focusing on making significant improvements to our Android app, including bug fixes, new features, and enhancements to overall performance.

We will notify you when the new version of our app is ready, and the new repository will be open to accept contributions. Thank you for your understanding, we appreciate your patience and support as we work to create a better experience.
